File: GpuData/Core+PCache1/11.PCache1+PageAllocation.cs
using System;
using Pgno = System.UInt32;
using System.Diagnostics;
using System.Text;
namespace Core
{
    public partial class PCache1
    {
        private static void BufferSetup(object buffer, int size, int n)
        {
            if (pcache1.IsInit)
            {
                size = SysEx.ROUNDDOWN8(size);
                pcache1.SizeSlot = size;
                pcache1.Slots = pcache1.FreeSlots = n;
                pcache1.Reserves = (n > 90 ? 10 : (n / 10 + 1));
                pcache1.Start = buffer;
                pcache1.End = null;
                pcache1.Free = null;
                pcache1.UnderPressure = false;
                while (n-- > 0)
                {
                    var p = new PgFreeslot();
                    p.Next = pcache1.Free;
                    pcache1.Free = p;
                }
                pcache1.End = buffer;
            }
        }

        private static PgHdr Alloc(int bytes)
        {
            PgHdr p = null;
            Debug.Assert(MutexEx.NotHeld(pcache1.Mutex));
            StatusEx.StatusSet(StatusEx.STATUS.PAGECACHE_SIZE, bytes);
            if (bytes <= pcache1.SizeSlot)
            {
                MutexEx.Enter(pcache1.Mutex);
                p = pcache1.Free._PgHdr;
                if (p != null)
                {
                    pcache1.Free = pcache1.Free.Next;
                    pcache1.FreeSlots--;
                    pcache1.UnderPressure = (pcache1.FreeSlots < pcache1.Reserves);
                    Debug.Assert(pcache1.FreeSlots >= 0);
                    StatusEx.StatusAdd(StatusEx.STATUS.PAGECACHE_USED, 1);
                }
                MutexEx.Leave(pcache1.Mutex);
            }
            if (p == null)
            {
                // Memory is not available in the SQLITE_CONFIG_PAGECACHE pool.  Get it from sqlite3Malloc instead.
                p = new PgHdr();
                {
                    var sz = bytes;
                    MutexEx.Enter(pcache1.Mutex);
                    StatusEx.StatusAdd(StatusEx.STATUS.PAGECACHE_OVERFLOW, sz);
                    MutexEx.Leave(pcache1.Mutex);
                }
                SysEx.MemdebugSetType(p, SysEx.MEMTYPE.PCACHE);
            }
            return p;
        }

        private static int Free(ref PgHdr p)
        {
            int freed = 0;
            if (p == null)
                return 0;
            if (p.CacheAllocated)
            {
                MutexEx.Enter(pcache1.Mutex);
                StatusEx.StatusAdd(StatusEx.STATUS.PAGECACHE_USED, -1);
                var slot = new PgFreeslot(p);
                slot.Next = pcache1.Free;
                pcache1.Free = slot;
                pcache1.FreeSlots++;
                pcache1.UnderPressure = (pcache1.FreeSlots < pcache1.Reserves);
                Debug.Assert(pcache1.FreeSlots <= pcache1.Slots);
                MutexEx.Leave(pcache1.Mutex);
            }
            else
            {
                Debug.Assert(SysEx.MemdebugHasType(p, SysEx.MEMTYPE.PCACHE));
                SysEx.MemdebugSetType(p, SysEx.MEMTYPE.HEAP);
                freed = SysEx.MallocSize(p._Data);
                MutexEx.Enter(pcache1.Mutex);
                StatusEx.StatusAdd(StatusEx.STATUS.PAGECACHE_OVERFLOW, -freed);
                MutexEx.Leave(pcache1.Mutex);
                SysEx.Free(ref p._Data);
            }
            return freed;
        }

#if ENABLE_MEMORY_MANAGEMENT
        private static int MemSize(PgHdr p)
        {
            if (p.CacheAllocated)
                return pcache1.SizeSlot;
            else
            {
                Debug.Assert(SysEx.MemdebugHasType(p, SysEx.MEMTYPE.PCACHE));
                SysEx.MemdebugSetType(p, SysEx.MEMTYPE.HEAP);
                var size = SysEx.MallocSize(p._Data);
                SysEx.MemdebugSetType(p, SysEx.MEMTYPE.PCACHE);
                return size;
            }
        }
#endif

        private PgHdr1 AllocPage()
        {
            var pPg = Alloc(SizePage);
            var p = new PgHdr1();
            p.Page.Buffer = pPg;
            p.Page.Extra = null; // TODO: map for extra
            if (Purgeable)
                Group.CurrentPages++;
            return p;
        }

        private static void FreePage(ref PgHdr1 p)
        {
            if (SysEx.ALWAYS(p))
            {
                var cache = p.Cache;
                if (cache.Purgeable)
                    cache.Group.CurrentPages--;
                Free(ref p.Page);
            }
        }

        private static PgHdr PageMalloc(int sz) { return Alloc(sz); }

        public static void PageFree(ref byte[] p)
        {
            if (p != null)
            {
                SysEx.Free(ref p);
                p = null;
            }
        }
        public static void PageFree(ref PgHdr p) { Free(ref p); }

        private bool UnderMemoryPressure() { return (pcache1.Slots != 0 && SizePage <= pcache1.SizeSlot ? pcache1.UnderPressure : MallocEx.HeapNearlyFull()); }
    }
}

File: src/GpuData/Core+PCache1/11.PCache1+PageAllocation.cpp
namespace Core
{
	static void BufferSetup(void *buffer, int size, int n)
	{
		if (pcache1.IsInit)
		{
			size = SysEx::ROUNDDOWN8(size);
			pcache1.SizeSlot = size;
			pcache1.Slots = pcache1.FreeSlots = n;
			pcache1.nReserve = (n > 90 ? 10 : (n / 10 + 1));
			pcache1.Start = buffer;
			pcache1.Free = nullptr;
			pcache1.UnderPressure = false;
			while (n--)
			{
				PgFreeslot *p = (PgFreeslot *)buffer;
				p->Next = pcache1.Free;
				pcache1.Free = p;
				//TODO: change to buffer += size;
				buffer = (void *)&((char *)buffer)[size];
			}
			pcache1.End = buffer;
		}
	}

	static void *Alloc(int bytes)
	{
		assert(Mutex_NotHeld(pcache1.Group.Mutex));
		StatusEx::StatusSet(StatusEx::STATUS_PAGECACHE_SIZE, bytes);
		PgHdr *p = nullptr;
		if (bytes <= pcache1.SizeSlot)
		{
			MutexEx::Enter(pcache1.Mutex);
			p = (PgHdr1 *)pcache1.Free;
			if (p)
			{
				pcache1.Free = pcache1.Free->Next;
				pcache1.FreeSlots--;
				pcache1.UnderPressure = (pcache1.FreeSlots < pcache1.Reserves);
				assert(pcache1.FreeSlots >= 0);
				StatusEx::StatusAdd(StatusEx::STATUS_PAGECACHE_USED, 1);
			}
			MutexEx::Leave(pcache1.Mutex);
		}
		if (!p)
		{
			// Memory is not available in the SQLITE_CONFIG_PAGECACHE pool.  Get it from sqlite3Malloc instead.
			p = sqlite3Malloc(bytes);
#ifndef DISABLE_PAGECACHE_OVERFLOW_STATS
			if (p)
			{
				int sz = sqlite3MallocSize(p);
				MutexEx::Enter(pcache1.Mutex);
				StatusEx::StatusAdd(StatusEx::STATUS_PAGECACHE_OVERFLOW, sz);
				MutexEx::Leave(pcache1.Mutex);
			}
#endif
			SysEx::MemdebugSetType(p, SysEx::MEMTYPE_PCACHE);
		}
		return p;
	}

	static int Free(void *p)
	{
		int freed = 0;
		if (p == nullptr)
			return 0;
		if (p >= pcache1.Start && p < pcache1.End)
		{
			MutexEx::Enter(pcache1.Mutex);
			StatusEx::StatusAdd(StatusEx::STATUS_PAGECACHE_USED, -1);
			PgFreeslot *slot = (PgFreeslot *)p;
			slot->Next = pcache1.Free;
			pcache1.Free = slot;
			pcache1.FreeSlots++;
			pcache1.UnderPressure = (pcache1.FreeSlot < pcache1.Reserves);
			assert(pcache1.FreeSlots <= pcache1.Slots);
			MutexEx::Leave(pcache1.Mutex);
		}
		else
		{
			assert(SysEx::MemdebugHasType(p, SysEx::MEMTYPE_PCACHE));
			SysEx::MemdebugSetType(p, SysEx::MEMTYPE_HEAP);
			freed = SysEx::MallocSize(p);
#ifndef DISABLE_PAGECACHE_OVERFLOW_STATS
			MutexEx::Enter(pcache1.Mutex);
			StatusEx::StatusAdd(StatusEx::STATUS_PAGECACHE_OVERFLOW, -freed);
			Mutex::Leave(pcache1.Mutex);
#endif
			SysEx::Free(p);
		}
		return freed;
	}

#ifdef ENABLE_MEMORY_MANAGEMENT
	static int MemSize(void *p)
	{
		if (p >= pcache1.Start && p < pcache1.End)
			return pcache1.SizeSlot;
		assert(SysEx::MemdebugHasType(p, SysEx::MEMTYPE_PCACHE) );
		SysEx::MemdebugSetType(p, SysEx::MEMTYPE_HEAP);
		int size = SysEx::MallocSize(p);
		SysEx::MemdebugSetType(p, SysEx::MEMTYPE_PCACHE);
		return size;
	}
#endif

	static PgHdr1 *AllocPage(PCache1 *t)
	{
		// The group mutex must be released before pcache1Alloc() is called. This is because it may call sqlite3_release_memory(), which assumes that this mutex is not held.
		assert(MutexEx::Held(t->Group->Mutex));
		PCache1::LeaveMutex(t->Group);
		PgHdr1 *p = nullptr;
		void *pPg;
#ifdef PCACHE_SEPARATE_HEADER
		pPg = Alloc(t->SizePage);
		p = SysEx::Malloc(sizeof(PgHdr1) + t->SizeExtra);
		if (!pPg || !p)
		{
			Free(pPg);
SysEx:Free(p);
			pPg = nullptr;
		}
#else
		pPg = Alloc(sizeof(PgHdr1) + t->SizePage + t->SizeExtra);
		p = (PgHdr1 *)&((u8 *)pPg)[t->SizePage];
#endif
		PCache1::EnterMutex(t->Group);
		if (pPg)
		{
			p->Page.Buffer = pPg;
			p->Page.Extra = &p[1];
			if (t->Purgeable)
				t->Group->CurrentPages++;
			return p;
		}
		return nullptr;
	}

	static void FreePage(PgHdr1 *p)
	{
		if (SysEx::ALWAYS(p))
		{
			PCache1 *cache = p->Cache;
			assert(MutexEx::Held(p->Cache->Group->Mutex));
			Free(p->Page.Buffer);
#ifdef PCACHE_SEPARATE_HEADER
			SysEx::Free(p);
#endif
			if (cache->Purgeable)
				cache->Group->CurrentPages--;
		}
	}

	/*
	** Malloc function used by SQLite to obtain space from the buffer configured
	** using sqlite3_config(SQLITE_CONFIG_PAGECACHE) option. If no such buffer
	** exists, this function falls back to sqlite3Malloc().
	*/
	void *sqlite3PageMalloc(int sz){
		return pcache1Alloc(sz);
	}

	/*
	** Free an allocated buffer obtained from sqlite3PageMalloc().
	*/
	void sqlite3PageFree(void *p){
		pcache1Free(p);
	}


	/*
	** Return true if it desirable to avoid allocating a new page cache
	** entry.
	**
	** If memory was allocated specifically to the page cache using
	** SQLITE_CONFIG_PAGECACHE but that memory has all been used, then
	** it is desirable to avoid allocating a new page cache entry because
	** presumably SQLITE_CONFIG_PAGECACHE was suppose to be sufficient
	** for all page cache needs and we should not need to spill the
	** allocation onto the heap.
	**
	** Or, the heap is used for all page cache memory but the heap is
	** under memory pressure, then again it is desirable to avoid
	** allocating a new page cache entry in order to avoid stressing
	** the heap even further.
	*/
	static int pcache1UnderMemoryPressure(PCache1 *pCache){
		if( pcache1.nSlot && (pCache->szPage+pCache->szExtra)<=pcache1.szSlot ){
			return pcache1.bUnderPressure;
		}else{
			return sqlite3HeapNearlyFull();
		}
	}

}
